File: atcoder/past17-2/i.py
from collections import defaultdict

n=int(input())
maxl=5
s=[input().rstrip() for _ in range(n)]
h=[defaultdict(lambda: set()) for _ in range(maxl+1)]
for idx,si in enumerate(s):
    h[len(si)][si] |= {idx}

hh=[defaultdict(lambda: set()) for _ in range(maxl+1)]
ans=0
for l in range(maxl,0,-1):
    for k,v in h[l].items():
        ans+=len(v)**2
        ans+=len(v)*len(hh[l][k])
        for idx in range(l):
            t=k[:idx]+k[idx+1:]
            hh[l-1][t]|=v
    for k,v in hh[l].items():
        for idx in range(l):
            t=k[:idx]+k[idx+1:]
            hh[l-1][t]|=v

print(ans)

File: atcoder/past17-2/i.rb
require 'set'
n=gets.to_i
maxl=5
s=(1..n).map{gets.chomp}
h=(0..maxl).map{Hash.new{Set[]}}
s.each.with_index{|si,idx|
    h[si.size][si]+=Set[idx]
}
hh=(0..maxl).map{Hash.new{Set[]}}
ans=0
maxl.downto(1){|l|
#    p [ans,h,hh]
    p hh
    h[l].each{|k,v|
        ans+=v.size**2
        ans+=v.size*hh[l][k].size
        (0...l).each{|idx|
            t=k.dup
            t[idx]=""
            hh[l-1][t]+=v
        }
    }
    hh[l].each{|k,v|
        (0...l).each{|idx|
            t=k.dup
            t[idx]=""
            hh[l-1][t]+=v
        }
    }
}
#p [ans,h]
p ans